TESTS/mbed_hal_fpga_ci_test_shield/gpio/main.cpp

```diff
@@ -71,7 +71,12 @@ void test_basic_input_output(PinName pin)
     gpio_mode(&gpio, PullUp);
     tester.gpio_write(MbedTester::LogicalPinGPIO0, 0, false);
     wait_us(HI_Z_READ_DELAY_US);
+    // Some targets don't support pull-up/pull-down mode.
+#if !defined(TARGET_NANO100) &&         \
+    !defined(TARGET_NUC472) &&          \
+    !defined(TARGET_M451)
```

Comment:
How can we eliminate these if target. Is this functionality a must ? Can we check differently if this mode is supported?

Reply:
Check `gpio_mode(...)` and `pin_mode(...)` but find no spec on the `mode` parameter. It seems most targets support `PullDown`/`PullUp` input pull mode. However, old-chip based Nuvoton targets like above don't support it. And there's no configuration option to disable it.

```
void gpio_mode(gpio_t *obj, PinMode mode);
void pin_mode(PinName pin, PinMode mode);
```

Reply:
@ARMmbed/mbed-os-hal Does this look correct as it is?

Reply:
@ccli8 The whole `// Test input, pull-up mode.` should be excluded if these targets do not support pull-up mode.

Reply:
> How can we eliminate these if target. Is this functionality a must ? Can we check differently if this mode is supported?
> 
> Having this ifdef is not ideal, means we are testing something is not either supported by all targets or we are missing a check to ask for "is this supported, test it".

Frankly, I didn't expect any target that does not support pull-up/pull-down modes. The only thing that comes to my mind, to make this a bit more readable, is to define a new macro in *targets.json* file. That could be either `"GPIO_PINMODE"` in `"device_has"` section or `"GPIO_PINMODE_NOT_SUPPORTED"` in `"macros"`. @ccli8, @0xc0170, @OPpuolitaival, what do you think?

Reply:
Before that, I think the `mode` parameter should be well defined. It just allows `PullNone`/`PullUp`/`PullDown`, or it can also allow other modes like target-specific? Only meaningful for input mode?

```C
void gpio_mode(gpio_t *obj, PinMode mode);
void pin_mode(PinName pin, PinMode mode);
```